src/ir/types/type_factory.cc

```diff
@@ -0,0 +1,44 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#include "src/ir/types/type_factory.h"
+
+#include <memory>
+
+#include "src/ir/types/entity_type.h"
+#include "src/ir/types/primitive_type.h"
+#include "src/ir/types/schema.h"
+
+namespace raksha::ir::types {
+
+Type TypeFactory::MakePrimitiveType() {
+  return Type(std::make_unique<PrimitiveType>());
+}
+
+Type TypeFactory::MakeEntityType(const Schema& schema) {
+  return Type(std::make_unique<EntityType>(schema));
+}
```

Comment:
This pattern means that every use of a `PrimitiveType` must use a different instance of that `PrimitiveType`, as each usage point owns its `PrimitiveType`. As mentioned in the other PR, perhaps the factory should return a non-owning ID for the type which can be used to look up the full type, which would be owned by the factory.

Reply:
Yes, the plan is to canonicalize all types include `PrimitiveType`. As I mentioned in the other PR, I wanted to introduce the `Type` wrapper and `TypeFactory` so that we can start building out the `Operator`, `Operation`, etc. Do you see any concerns? e.g., does the `Type` wrapper as defined prevent us from achieving the canonicalization goals?